File: video-app/android-native/app/src/main/java/com/videoapp/player/util/NetworkUtils.java
package com.videoapp.player.util;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.Network;
import android.net.NetworkCapabilities;
import android.net.NetworkInfo;
import android.os.Build;
import android.util.Log;

import androidx.annotation.NonNull;

/**
 * Network utilities for checking connectivity
 */
public final class NetworkUtils {

    private static final String TAG = "NetworkUtils";

    private NetworkUtils() {
        // Private constructor to prevent instantiation
    }

    /**
     * Check if the device has an active network connection
     * Handles different API levels properly for Android 9+ compatibility
     */
    public static boolean isNetworkAvailable(@NonNull Context context) {
        try {
            ConnectivityManager connectivityManager =
                    (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
            if (connectivityManager == null) {
                return false;
            }

            Network network = connectivityManager.getActiveNetwork();
            if (network == null) {
                // Fallback for older behavior
                return isNetworkAvailableFallback(connectivityManager);
            }

            NetworkCapabilities capabilities = connectivityManager.getNetworkCapabilities(network);
            if (capabilities == null) {
                // Fallback if capabilities are null
                return isNetworkAvailableFallback(connectivityManager);
            }

            // Check for internet capability
            // Note: NET_CAPABILITY_VALIDATED might not be set on some networks,
            // so we also check for basic connectivity
            boolean hasInternet = capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET);
            boolean isValidated = capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED);

            // On Android 9 (API 28), some networks might not have VALIDATED flag set properly
            // So we accept if at least INTERNET capability is present
            if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.P) {
                // For Android 9 and below, be more lenient
                return hasInternet && (isValidated ||
                        capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) ||
                        capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR));
            }

            return hasInternet && isValidated;
        } catch (Exception e) {
            Log.e(TAG, "Error checking network availability", e);
            // In case of error, return false to be safe - let the calling code handle it
            return false;
        }
    }

    /**
     * Fallback network check using deprecated API for maximum compatibility
     */
    @SuppressWarnings("deprecation")
    private static boolean isNetworkAvailableFallback(@NonNull ConnectivityManager connectivityManager) {
        try {
            NetworkInfo activeNetworkInfo = connectivityManager.getActiveNetworkInfo();
            return activeNetworkInfo != null && activeNetworkInfo.isConnected();
        } catch (Exception e) {
            Log.e(TAG, "Error in fallback network check", e);
            return false; // Return false if we can't determine connectivity
        }
    }

    /**
     * Check if connected to WiFi
     */
    public static boolean isWifiConnected(@NonNull Context context) {
        try {
            ConnectivityManager connectivityManager =
                    (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
            if (connectivityManager == null) {
                return false;
            }

            Network network = connectivityManager.getActiveNetwork();
            if (network == null) {
                return false;
            }

            NetworkCapabilities capabilities = connectivityManager.getNetworkCapabilities(network);
            if (capabilities == null) {
                return false;
            }

            return capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI);
        } catch (Exception e) {
            Log.e(TAG, "Error checking WiFi connection", e);
            return false;
        }
    }

    /**
     * Get friendly network error message
     */
    @NonNull
    public static String getNetworkErrorMessage(@NonNull Context context) {
        if (!isNetworkAvailable(context)) {
            return "网络连接不可用，请检查网络设置";
        } else {
            return "网络请求失败，请稍后重试";
        }
    }
}

File: video-app/android-native/app/src/main/java/com/videoapp/player/util/NetworkUtils.kt
package com.videoapp.player.util

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities

/**
 * Network utilities for checking connectivity
 */
object NetworkUtils {
    
    /**
     * Check if the device has an active network connection
     */
    fun isNetworkAvailable(context: Context): Boolean {
        val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
            ?: return false
        
        val network = connectivityManager.activeNetwork ?: return false
        val capabilities = connectivityManager.getNetworkCapabilities(network) ?: return false
        
        return capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET) &&
               capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED)
    }
    
    /**
     * Check if connected to WiFi
     */
    fun isWifiConnected(context: Context): Boolean {
        val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
            ?: return false
        
        val network = connectivityManager.activeNetwork ?: return false
        val capabilities = connectivityManager.getNetworkCapabilities(network) ?: return false
        
        return capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI)
    }
    
    /**
     * Get friendly network error message
     */
    fun getNetworkErrorMessage(context: Context): String {
        return if (!isNetworkAvailable(context)) {
            "网络连接不可用，请检查网络设置"
        } else {
            "网络请求失败，请稍后重试"
        }
    }
}
